packages/doenetml-worker-rust/doenetml-core/src/parse_json.rs

```diff
@@ -0,0 +1,76 @@
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+use crate::{state::StateVarValue, Action};
+
+/// The structure of the json argument in a call to handle_action.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ActionStructure {
+    component_idx: usize,
+    action_name: String,
+    args: HashMap<String, ArgValue>,
+}
+
+/// Each value in an action args must be a quantity that can be converted
+/// into a `StateVarValue` or a vector of `StateVarValue`.
+///
+/// Currently just booleans, Numbers, arrays of numbers, and strings are supported.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(untagged)]
+enum ArgValue {
+    Bool(bool),
+    Number(serde_json::Number),
+    NumberArray(Vec<serde_json::Number>),
+    String(String),
+}
+
+/// Parse the action json, converting all fields of args into a vector of StateVarVal.
+pub fn parse_action_from_json(action: &str) -> Result<Action, String> {
+    let action_structure: ActionStructure =
+        serde_json::from_str(action).map_err(|e| e.to_string())?;
+
+    let component_idx = action_structure.component_idx;
+    let action_name = action_structure.action_name.clone();
+    let args: HashMap<String, Vec<StateVarValue>> = action_structure
+        .args
+        .into_iter()
+        .map(|(k, v)| (k, v.into()))
+        .collect();
+
+    Ok(Action {
+        component_idx,
+        action_name,
+        args,
+    })
+}
+
+impl From<serde_json::Number> for StateVarValue {
+    /// A json number will be converted to an integer if it is an integer.
+    ///
+    /// For this reason, it's important that we can convert StateVarValue::Integer to f64,
```

Comment:
Are there cases where we'd want to use `BigInt` instead?

Reply:
I haven't needed a BigInt yet, but I could imagine eventually folks will want it. I'm trying to build a structure where we can easily add additional data types.